server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java

```diff
@@ -0,0 +1,208 @@
+        if (family.equals(DataFileColumnFamily.NAME)) {
+          upgradeDataFileCF(k, v, batchWriter, tableName);
+        } else if (family.equals(ChoppedColumnFamily.NAME)) {
+          removeChoppedCF(k, batchWriter, tableName);
+        } else if (family.equals(ExternalCompactionColumnFamily.NAME)) {
+          removeExternalCompactionCF(k, batchWriter, tableName);
+        } else {
+          log.warn("Received unexpected column family processing references: " + family);
+        }
+      });
+    } catch (MutationsRejectedException mex) {
+      log.warn("Failed to update reference for table: " + tableName);
+      log.warn("Constraint violations: {}", mex.getConstraintViolationSummaries());
+      throw new IllegalStateException("Failed to process table: " + tableName, mex);
+    } catch (Exception ex) {
+      throw new IllegalStateException("Failed to process table: " + tableName, ex);
+    }
+  }
+
+  @VisibleForTesting
+  void upgradeDataFileCF(final Key key, final Value value, final BatchWriter batchWriter,
+      final String tableName) {
+    String file = key.getColumnQualifier().toString();
+    // filter out references if they are in the correct format already.
+    if (fileNeedsConversion(file)) {
+      var fileJson = StoredTabletFile.of(new Path(file)).getMetadataText();
+      try {
+        Mutation update = new Mutation(key.getRow());
+        update.at().family(DataFileColumnFamily.STR_NAME).qualifier(fileJson).put(value);
+        log.trace("table: {}, adding: {}", tableName, update.prettyPrint());
+        batchWriter.addMutation(update);
+
+        Mutation delete = new Mutation(key.getRow());
+        delete.at().family(DataFileColumnFamily.STR_NAME).qualifier(file).delete();
+        log.trace("table {}: deleting: {}", tableName, delete.prettyPrint());
+        batchWriter.addMutation(delete);
+      } catch (MutationsRejectedException ex) {
+        // include constraint violation info in log - but stop upgrade
+        log.warn(
+            "Failed to update file reference for table: " + tableName + ". row: " + key.getRow());
+        log.warn("Constraint violations: {}", ex.getConstraintViolationSummaries());
+        throw new IllegalStateException("File conversion failed. Aborting upgrade", ex);
+      }
+    }
+  }
+
+  @VisibleForTesting
+  void removeChoppedCF(final Key key, final BatchWriter batchWriter, final String tableName) {
+    Mutation delete = null;
+    try {
+      delete = new Mutation(key.getRow()).at().family(ChoppedColumnFamily.STR_NAME)
+          .qualifier(ChoppedColumnFamily.STR_NAME).delete();
+      log.warn(
+          "Deleting chopped reference from:{}. Previous split or delete may not have completed cleanly. Ref: {}",
+          tableName, delete.prettyPrint());
+      batchWriter.addMutation(delete);
+    } catch (MutationsRejectedException ex) {
+      log.warn("Failed to delete obsolete chopped CF reference for table: " + tableName + ". Ref: "
```

Comment:
Need to rethrow this exception.  The batch writer queues up mutations and does not send them immediately.  When addMuation throws an exception it may not have been caused by the most recently added mutation.  Could be from a mutation for changing a file.

Reply:
Added in bceac27d1c